lib/src/components/NxDrawer/NxDrawer.tsx

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2019-present Sonatype, Inc.
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/.
+ */
+import React, {
+  HTMLAttributes,
+  ReactNode,
+  useContext,
+  useEffect,
+  useRef,
+  useState
+} from 'react';
+import classnames from 'classnames';
+
+import AbstractDialog from '../AbstractDialog/AbstractDialog';
+import NxCloseButton from '../NxCloseButton/NxCloseButton';
+import withClass from '../../util/withClass';
+
+import { Props, propTypes } from './types';
+
+import './NxDrawer.scss';
+
+interface NxDrawerHeaderProps extends HTMLAttributes<HTMLElement>{
+  children: ReactNode;
+}
+
+interface NxDrawerContextValue {
+  closeDrawer: () => void;
+  open: boolean;
+}
+
+const NxDrawerContext = React.createContext<NxDrawerContextValue>({
+  closeDrawer: () => {},
+  open: false
+});
+
+const NxDrawerHeader = (props: NxDrawerHeaderProps) => {
+  const {
+    className,
+    children,
+    ...attrs
+  } = props;
+
+  const closeButtonRef = useRef<HTMLButtonElement>(null);
+
+  const { closeDrawer } = useContext(NxDrawerContext);
+
+  const classes = classnames('nx-drawer-header', className);
+
+  return (
+    <header className={classes} role="none" {...attrs}>
+      <NxCloseButton className="nx-drawer-header__cancel-button"
+                     type="button"
+                     ref={closeButtonRef}
+                     onClick={() => closeDrawer()}>
+        Close
+      </NxCloseButton>
+      {children}
+    </header>
+  );
+};
+
+function NxDrawer(props: Props) {
+  const {
+    className,
+    open,
+    onClose,
+    onCancel,
+    children,
+    variant,
+    ...attrs
+  } = props;
+
+  // Whether the drawer is _fully_ closed (i.e. not open and not animating between open/closed)
+  const [isClosed, setIsClosed] = useState(!open);
+
+  const dialogRef = useRef<HTMLDialogElement>(null);
+
+  const closeDrawer = () => onClose();
+
+  const handleAnimationEnd = () => {
+    if (!open) {
+      setIsClosed(true);
+
+      if (onCancel) {
+        onCancel();
+      }
+    }
+  };
+
+  useEffect(() => {
+    if (open) {
+      setIsClosed(false);
+
+      const listener = (event: MouseEvent) => {
+        if (dialogRef.current && !dialogRef.current.contains(event.target as Node)) {
+          closeDrawer();
+        }
+      };
+
+      document.addEventListener('click', listener);
+
+      return () => document.removeEventListener('click', listener);
+    }
+    else {
```

Comment:
Hi @rpokorny, just wanted to know the reasoning behind returning undefined here and there were also a few others in `AbstractDialog.

Reply:
> > Hi @rpokorny, just wanted to know the reasoning behind returning undefined here and there were also a few others in `AbstractDialog.
> 
> It's only to make TypeScript happy. Without it you get "Not all code paths return a value". But yes functionally it's equivalent to having no return statement at all on these code paths.

Ah I see it now, thanks!